### PR TITLE
Collecting commits without Docker

### PR DESCRIPTION
**Pull request by ninjinkun**
- Add `--executable` option to the `commit` command
  - It has `jar` and `docker` options. `jar` is the default.
- Download `exec_deploy.jar` if it doesn't exist

## What point do I want you to review?
- I added versioning to `exec_deploy.jar` on S3 and local. Do you agree it works well?
- I added `~/.launchable` as a jar cache directory. Do you think is this directory right place to store jar files?

### CODE REVIEW

File: launchable/utils/ingester_jar.py
Comment (ninjinkun) [inline launchable/utils/ingester_jar.py:1]: If `exe_deploy.jar` is updated, you should bump up this version number. I assume the number will be the same as the CLI version number.

File: launchable/commands/record/commit.py
Comment (ninjinkun) [inline launchable/commands/record/commit.py:12]: I specified `~/.launchable/jar/ingester/0.1.0/exe_deploy.jar` as a local jar file path. Is this okay?

**Comment by ninjinkun** [comment]
@kohsuke Please review this PR. Its run result is here https://github.com/launchableinc/rocket-car/runs/1353609834

File: launchable/commands/record/commit.py
Comment (kohsuke) [inline launchable/commands/record/commit.py:40]: I could be wrong, but [according to this](https://requests.readthedocs.io/en/latest/user/quickstart/#errors-and-exceptions) it seems to me like `res.raise_for_status()` would be needed to detect 404 and other errors.

File: launchable/commands/record/commit.py
Comment (kohsuke) [inline launchable/commands/record/commit.py:41]: I'm also mildly concerned that every invocation of build recording now creates a request to S3 and any problems there will render the entire command inoperable. We already know for example that BMW's build environment needs IP-based whitelisting, so I think this would create an extra step for them.

If we download a file into a temporary file, then rename it to the expected name, then `os.path.exists(...)` would be sufficient to verify that completeness of a file, so that would work avoid this problem. In this way, BMW can also just manually download a file and place it in the location they expect. Also handy for people who create a builder image.

File: launchable/commands/record/commit.py
Comment (kohsuke) [inline launchable/commands/record/commit.py:53]: For interactive use this status report is nice but for CI systems where this command is expected to be used, this is effectively no different from NL, so it'd be quite noisy.

Not a major point, but I'd suggest toning this down. For example, one line to report the start of download (some CI systems can only show a line after it sees a whole line), then one dot for every 10MB, and NL at the end of download, would be nice. Alternatively, you can check if the stdout is a terminal, and use that as a cue to produce output geared for humans like this.

(But really, the whole on demand download is an extra nice-to-have, so it's probably not wise to get too smart in that regard.)

File: launchable/commands/record/commit.py
Comment (kohsuke) [inline launchable/commands/record/commit.py:75]: When the commit ingester fails, how should we handle it? It seems to me that the code today will silently ignore that and moves on. Perhaps that's intended (so as not to disrupt the customer's build/test), or maybe it's an oversight.

More broadly, this does seem like something we should think carefully. There are competing factors we need to balance -- we don't want to disrupt customers' build & test and get kicked out of their build process, but we also want them to notice if things aren't working as it's supposed to be.

Whatever our policy is that we define there, we should apply that consistently in other commands, too.

File: launchable/commands/record/commit.py
Comment (kohsuke) [inline launchable/commands/record/commit.py:70]: Nice touch!

**Review by kohsuke — APPROVED** [review]
I'm marking this as approved so as not to block the delivery to the rocket-car company, which is on the critical path for the company.

I left some comments around the design of the capabilities.

File: launchable/commands/record/commit.py
Comment (ninjinkun) [reply launchable/commands/record/commit.py:40]: Got it. I'm going to implement to handle HTTP errors.

File: launchable/commands/record/commit.py
Comment (ninjinkun) [reply launchable/commands/record/commit.py]: Thank you for your suggestion. I implemented moving `jar` from the temporary directory to the cache directory. As a result, the CLI doesn't need to access S3 if the jar file was downloaded.

File: launchable/commands/record/commit.py
Comment (ninjinkun) [reply launchable/commands/record/commit.py]: Okay. I fixed to tone down this prompt like this
```
Downloading exe_deploy.jar (262.03 MB)
....
```

File: launchable/commands/record/commit.py
Comment (ninjinkun) [reply launchable/commands/record/commit.py:88]: It's simply an oversight. I think `LAUNCHABLE_REPORT_ERROR` environment variable is a good option for this fail.  I'll fix it at another PR after merge this PR.

**Comment by ninjinkun** [comment]
@kohsuke Thank you for your review! I merge it before your second review because I want to roll out it to the rocket car company. If you have another review, please leave comments on this PR. I'll fix them on another PR.

**Pull request merged**
